packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -913,15 +913,25 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
                     return create404Response(request);
                 }
             } else if (requestParser.pathParts.length > 0) {
-                const wait =
-                    typeof request.headers?.wait === "boolean" ? request.headers.wait : undefined;
-
-                const dataStore = await this.getDataStore(id, wait);
-                const subRequest = requestParser.createSubRequest(1);
-                // We always expect createSubRequest to include a leading slash, but asserting here to protect against
-                // unintentionally modifying the url if that changes.
-                assert(subRequest.url.startsWith("/"), "Expected createSubRequest url to include a leading slash");
-                return dataStore.IFluidRouter.request(subRequest);
+                /**
+                 * If the request has "errorOnUnreferencedResources" header, we need to return an error if the data
+                 * store being requested is marked as unreferenced as per the data store's initial summary.
+                 *
+                 * This is a workaround to handle scenarios where a data store shared with an external app is deleted
+                 * and marked as unreferenced by GC. Returning an error will fail to load the data store for the app.
+                 */
+                const wait = typeof request.headers?.wait === "boolean" ? request.headers.wait : undefined;
+                const dataStore = request.headers?.errorOnUnreferencedResources
+                    ? await this.getDataStoreIfInitiallyReferenced(id, wait)
```

Comment:
I'm not sure I like the fact that we encode that behavior differences are only at data store layer.
I.e. DataObject might create DDS per some conditions and expose it via request route, accessible to external users (note that DDS itself should never be exposed / should not be API surface, but nothing prevents DataObject to expose it via proxy). Then same rules RE GC may apply to this route / object and it might be marked unreferenced and deleted in the future.

Reply:
Such requests will also have to go via this path right?
For unreferenced data stores, I would like the behavior to be the same as it would be when the data store is actually deleted. In that case, we would not have the data store context at all and so the request would just stop at this layer.

Reply:
I think it's fine as it is. I was thinking more to incorporate it directly into getDataStore as a flag.

Reply:
Let's keep this code for now. We can incorporate into getDataStore if required later.